reports-backend/src/main/java/org/jboss/da/reports/impl/ReportsGeneratorImpl.java

```diff
@@ -81,35 +88,198 @@
+        updateArtifact(report, whiteArtifactsAndOrigin);
+    }
+
+    private void updateArtifact(ArtifactReport ar,
+            Map<WhiteArtifact, Set<ProductVersion>> allArtifacts) {
+
+        Set<ArtifactReport> dependencies = ar.getDependencies();
+
+        // Process the current artifact
+
+        GAV currentGav = ar.getGav();
+
+        List<String> availableVersionsFromWhitelist = getAvailableWhitelistVersions(allArtifacts,
+                currentGav);
+        Optional<String> bestVersion = getBestMatchVersionFromWhitelist(ar.getBestMatchVersion(),
+                currentGav, availableVersionsFromWhitelist);
+
+        if (ar.getAvailableVersions() != null)
+            availableVersionsFromWhitelist.addAll(ar.getAvailableVersions());
+
+        // Update available versions
+        ar.setAvailableVersions(availableVersionsFromWhitelist);
+
+        // Update best match version
+        ar.setBestMatchVersion(bestVersion);
+
+        // Recurse or end
+        if (dependencies == null || dependencies.isEmpty()) {
+            return;
+        } else {
+            for (ArtifactReport depArtifact : dependencies) {
+                updateArtifact(depArtifact, allArtifacts);
+            }
+        }
+    }
+
+    private Optional<String> getBestMatchVersionFromWhitelist(Optional<String> currentBestVersion,
+            GAV currentGav, List<String> availableVersionsFromWhitelist) {
+        // Establish best match version
+        Optional<String> bestWhiteVersion = versionFinder.getBestMatchVersionFor(currentGav,
+                availableVersionsFromWhitelist);
+        if (bestWhiteVersion != null && bestWhiteVersion.isPresent()) {
+            if (currentBestVersion.isPresent()) {
+                currentBestVersion = versionFinder.getBestMatchVersionFor(currentGav,
+                        Arrays.asList(bestWhiteVersion.get(), currentBestVersion.get()));
+
+            } else {
+                currentBestVersion = bestWhiteVersion;
+            }
+        }
+        return currentBestVersion;
+    }
+
+    private List<String> getAvailableWhitelistVersions(Map<WhiteArtifact, Set<ProductVersion>> allArtifacts,
+            GAV currentGav) {
+        Set<WhiteArtifact> whiteArtifactsForCurrentGav = allArtifacts.keySet().stream()
+                .filter(x -> {
+                    return x.getGa().getArtifactId()
+                            .equals(
+                           currentGav.getGA().getArtifactId())
+                            
+                            &&
+                           
+                           x.getGa().getGroupId()
+                            .equals(
+                           currentGav.getGA().getGroupId());
+                })
+                .collect(Collectors.toSet());
+                
+        List<String> whiteVersionsForCurrentGav = whiteArtifactsForCurrentGav.stream()
+                .map(x -> x.getVersion())
+                .collect(Collectors.toList());
```

Comment:
You can skip the lines 212-214 and use the map() on the filtered WhiteArtifact stream directly.